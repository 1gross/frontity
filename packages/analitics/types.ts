import { Package, Action } from "frontity/types";

export type Pageview = {
  url: string;
  title: string;
};

export type Event = {
  type: string;
  content: string;
};

interface Analytics extends Package {
  actions: {
    analytics: {
      afterCSR: Action<Analytics>;
      sendPageview: Action<Analytics, Pageview>;
      sendEvent: Action<Analytics, Event>;
    };
    [key: string]: {
      sendPageview: Action<Analytics, Pageview>;
      sendEvent: Action<Analytics, Event>;
    };
  };
  state: {
    analytics: {
      namespaces: string[];
    };
  };
}

export default Analytics;
