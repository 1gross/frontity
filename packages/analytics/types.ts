import { Package, Action, Connect } from "frontity/types";
import Source from "@frontity/source/types";

export type Pageview = {
  page: string;
  title: string;
};

export type Event = {
  category: string;
  action: string;
  label?: string;
};

interface Analytics extends Package {
  roots: {
    analytics: React.FC;
  };
  actions: {
    analytics: {
      sendPageview: Action<Analytics, Pageview>;
      sendEvent: Action<Analytics, Event>;
    };
    [key: string]: {
      sendPageview: Action<Analytics, Pageview>;
      sendEvent: Action<Analytics, Event>;
    };
  };
  state: {
    analytics: {
      namespaces: string[];
    };
    router?: {
      link: string;
    };
    source?: {
      get: Source["state"]["source"]["get"];
    };
  };
}

export default Analytics;
