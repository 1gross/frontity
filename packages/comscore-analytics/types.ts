import { ReactType } from "react";
import { Action, Package } from "frontity/types";
import Analytics, { Pageview } from "@frontity/analytics/types";

declare global {
  interface Window {
    COMSCORE: {
      beacon: Function;
    };
    _comscore: Array<{
      c1: string;
      c2: string;
    }>;
  }
}

/**
 * `@frontity/comscore-analytics` package.
 */
interface ComscoreAnalytics extends Package {
  roots: Analytics["roots"] & {
    comscoreAnalytics: ReactType;
  };
  state: Analytics["state"] & {
    /**
     * State properties for the Comscore analytics package.
     */
    comscoreAnalytics: {
      /**
       * Single tracking ID associated to a Comscore client.
       */
      trackingId?: string;
      /**
       * Array of tracking IDs associated to a Comscore client.
       */
      trackingIds?: string[];
    };
    analytics: {
      pageviews: {
        /**
         * Enable or disable sending pageviews to Comscore (default is `true`).
         */
        comscoreAnalytics: boolean;
      };
    };
  };
  actions: Analytics["actions"] & {
    /**
     * Actions from the Comscore analytics package.
     */
    comscoreAnalytics: {
      /**
       * Send a pageview to all Comscore traking ids defined in
       * `state.comscoreAnalytics.trackingIds`.
       *
       * @remarks
       * This action is called automatically if
       * `state.analytics.pageviews.comscoreAnalytics` is set to `true`.
       *
       * @param pageview - Object of type {@link Pageview}.
       */
      pageview: Action<ComscoreAnalytics, Pageview>;
    };
  };
}
export type Packages = ComscoreAnalytics & Analytics;

export default ComscoreAnalytics;
