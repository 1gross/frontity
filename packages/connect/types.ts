/**
 * The raw store object that contains state, actions an optionally libraries.
 * It's meant to be extended by the final user store.
 */
export interface StoreType {
  state: object;
  actions: object;
  libraries?: object;
}

/**
 * Interface for the options that can be passed to a function that creates proxies,
 * like observableState, mutableState, observableActions, executableActions...
 */
export interface Options {
  owner?: {
    type: "debug" | "action" | "component" | "when";
    name?: string;
  };
  mode?: "development" | "production";
}

/**
 * Helper to make all properties in Type readonly recursively.
 *
 * @param Type Any type.
 */
type RecursiveReadonly<Type> = {
  readonly [P in keyof Type]: Type[P] extends (...args: unknown[]) => unknown
    ? Type[P]
    : Type[P] extends object
    ? RecursiveReadonly<Type[P]>
    : Type[P]
};

/**
 * The observable state object of the store. It substitues the derived state functions for its return type.
 * It's meant to be used when the state needs to be observed, like in components or when functions.
 *
 * For example, the derived state:
 *  - `const userLength = ({ state }) => state.users.length;` can be used like a property: `state.userLength`.
 *  - `const userName = ({ state }) => index => state.users[index].name;` can be called with only the `index` argument: `state.userName(3)`.
 *
 * @param Store A store object containing state, actions and libraries.
 */
export type ObservableState<Store extends StoreType> = RecursiveReadonly<
  IterableState<Store["state"]>
>;

/**
 * The mutable state object of the store. It substitues the derived state functions for its return type.
 * It's meant to be used when the state can be mutated, like in actions.
 *
 * TODO: Make derived functions readonly.
 *
 * For example, the derived state:
 *  - `const userLength = ({ state }) => state.users.length;` can be used like a property: `state.userLength`.
 *  - `const userName = ({ state }) => index => state.users[index].name;` can be called with only the `index` argument: `state.userName(3)`.
 *
 * @param Store A store object containing state, actions and libraries.
 */
export type MutableState<Store extends StoreType> = IterableState<
  Store["state"]
>;

/**
 * Helper to iterate recursively after receiving the state.
 *
 * @param State The state object.
 */
type IterableState<State extends object> = {
  [P in keyof State]: State[P] extends (...args: unknown[]) => unknown
    ? ReturnType<State[P]>
    : State[P] extends object
    ? IterableState<State[P]>
    : State[P]
};

/**
 * The observable actions object of the store. It removes the need to pass the
 * argument of the first function. Observable actions should not be called.
 * It's meant to be used when the actions need to be only observed, like in when functions.
 *
 * TODO: Decide what to return instead of the function, maybe a boolean or an object
 * cotaining { args, name }.
 *
 * @param Store A store object containing state, actions and libraries.
 */
export type ObservableActions<Store extends StoreType> = RecursiveReadonly<
  IterableObservableActions<Store["actions"]>
>;

/**
 * Helper to iterate recursively after receiving the actions.
 *
 * @param Actions The actions object.
 */
type IterableObservableActions<Actions extends object> = {
  [P in keyof Actions]: Actions[P] extends (store: StoreType) => unknown
    ? boolean
    : Actions[P] extends object
    ? IterableObservableActions<Actions[P]>
    : never
};

/**
 * The executable actions of the store. It removes the need to pass the argument of the first function.
 * It's meant to be used when the actions need to be executed, like in components or other actions.
 *
 * @param Store A store object containing state, actions and libraries.
 */
export type ExecutableActions<Store extends StoreType> = RecursiveReadonly<
  IterableExecutableActions<Store["actions"]>
>;

/**
 * Helper to iterate recursively after receiving the actions.
 *
 * @param Actions The actions object.
 */
type IterableExecutableActions<Actions extends object> = {
  [P in keyof Actions]: Actions[P] extends (
    store: StoreType
  ) => (...args: infer Args) => Promise<void>
    ? (...args: Args) => Promise<void>
    : Actions[P] extends (store: StoreType) => (...args: infer Args) => void
    ? (...args: Args) => void
    : Actions[P] extends (store: StoreType) => void
    ? () => ReturnType<Actions[P]>
    : Actions[P] extends object
    ? IterableExecutableActions<Actions[P]>
    : never
};

/**
 * The store injected in the when functions.
 *
 * @param Store
 * The store containing the state, actions and libraries.
 */
type WhenStore<Store extends StoreType> = {
  state: ObservableState<Store>;
  actions: ObservableActions<Store>;
};

/**
 * The function passed to when. It receives the observable state and actions.
 *
 * @param Store
 * The store containing the state, actions and libraries.
 * @param Return
 * The return value of the when function.
 */
export interface When<Store extends StoreType, Return> {
  (store: WhenStore<Store>): Return;
}

/**
 * The store passed to actions, with mutable state, executable actions,
 * libraries and the when function.
 * It's meant to be used inside actions.
 *
 * @param Store
 * The store containing the state, actions and libraries.
 */
type ActionsStore<Store extends StoreType> = {
  state: MutableState<Store>;
  actions: ExecutableActions<Store>;
  libraries: Store["libraries"];
  when: <Store extends StoreType, Return>(
    observer: When<Store, Return>
  ) => Promise<Return>;
};

/**
 * An action function. It uses the store as the input of the first functions and the
 * rest of the inputs as the arguments of the second function.
 * It's meant to be used when defining the action.
 *
 * Examples:
 *  - `Action<YourStore>` ---> `({ state }) => { state.user = "Jon"; }`
 *  - `Action<YourStore, string>` ---> `({ state }) => name => { state.user = name; }`
 *  - `Action<YourStore, number, string>` ---> `({ state }) => (id, name) => { state.users[id].name = name; }`
 *
 * @param Store
 * The store containing the state, actions and libraries.
 * @param ArgumentX
 * The arguments of the action function. Supports up to 10 arguments.
 */
export type Action<
  Store extends StoreType,
  Argument1 = null,
  Argument2 = null,
  Argument3 = null,
  Argument4 = null,
  Argument5 = null,
  Argument6 = null,
  Argument7 = null,
  Argument8 = null,
  Argument9 = null,
  Argument10 = null
> = [Argument1] extends [null]
  ? (store: ActionsStore<Store>) => void
  : [Argument2] extends [null]
  ? (store: ActionsStore<Store>) => (arg1: Argument1) => void
  : [Argument3] extends [null]
  ? (store: ActionsStore<Store>) => (arg1: Argument1, arg2: Argument2) => void
  : [Argument4] extends [null]
  ? (
      store: ActionsStore<Store>
    ) => (arg1: Argument1, arg2: Argument2, arg3: Argument3) => void
  : [Argument5] extends [null]
  ? (
      store: ActionsStore<Store>
    ) => (
      arg1: Argument1,
      arg2: Argument2,
      arg3: Argument3,
      arg4: Argument4
    ) => void
  : [Argument6] extends [null]
  ? (
      store: ActionsStore<Store>
    ) => (
      arg1: Argument1,
      arg2: Argument2,
      arg3: Argument3,
      arg4: Argument4,
      arg5: Argument5
    ) => void
  : [Argument7] extends [null]
  ? (
      store: ActionsStore<Store>
    ) => (
      arg1: Argument1,
      arg2: Argument2,
      arg3: Argument3,
      arg4: Argument4,
      arg5: Argument5,
      arg6: Argument6
    ) => void
  : [Argument8] extends [null]
  ? (
      store: ActionsStore<Store>
    ) => (
      arg1: Argument1,
      arg2: Argument2,
      arg3: Argument3,
      arg4: Argument4,
      arg5: Argument5,
      arg6: Argument6,
      arg7: Argument7
    ) => void
  : [Argument9] extends [null]
  ? (
      store: ActionsStore<Store>
    ) => (
      arg1: Argument1,
      arg2: Argument2,
      arg3: Argument3,
      arg4: Argument4,
      arg5: Argument5,
      arg6: Argument6,
      arg7: Argument7,
      arg8: Argument8
    ) => void
  : [Argument10] extends [null]
  ? (
      store: ActionsStore<Store>
    ) => (
      arg1: Argument1,
      arg2: Argument2,
      arg3: Argument3,
      arg4: Argument4,
      arg5: Argument5,
      arg6: Argument6,
      arg7: Argument7,
      arg8: Argument8,
      arg9: Argument9
    ) => void
  : (
      store: ActionsStore<Store>
    ) => (
      arg1: Argument1,
      arg2: Argument2,
      arg3: Argument3,
      arg4: Argument4,
      arg5: Argument5,
      arg6: Argument6,
      arg7: Argument7,
      arg8: Argument8,
      arg9: Argument9,
      arg10: Argument10
    ) => void;

/**
 * An async action function. It uses the store as the input of the first functions and the
 * rest of the inputs as the arguments of the second function.
 * It's meant to be used when defining the async action.
 *
 * Examples:
 *  - `AsyncAction<YourStore>` ---> `async ({ state }) => { state.user = "Jon"; }`
 *  - `AsyncAction<YourStore, string>` ---> `({ state }) => async name => { state.user = name; }`
 *  - `AsyncAction<YourStore, number, string>` ---> `({ state }) => async (id, name) => { state.users[id].name = name; }`
 *
 * @param Store
 * The store containing the state, actions and libraries.
 * @param ArgumentX
 * The arguments of the action function. Supports up to 10 arguments.
 */
export type AsyncAction<
  Store extends StoreType,
  Argument1 = null,
  Argument2 = null,
  Argument3 = null,
  Argument4 = null,
  Argument5 = null,
  Argument6 = null,
  Argument7 = null,
  Argument8 = null,
  Argument9 = null,
  Argument10 = null
> = [Argument1] extends [null]
  ? (store: ActionsStore<Store>) => Promise<void>
  : [Argument2] extends [null]
  ? (store: ActionsStore<Store>) => (arg1: Argument1) => Promise<void>
  : [Argument3] extends [null]
  ? (
      store: ActionsStore<Store>
    ) => (arg1: Argument1, arg2: Argument2) => Promise<void>
  : [Argument4] extends [null]
  ? (
      store: ActionsStore<Store>
    ) => (arg1: Argument1, arg2: Argument2, arg3: Argument3) => Promise<void>
  : [Argument5] extends [null]
  ? (
      store: ActionsStore<Store>
    ) => (
      arg1: Argument1,
      arg2: Argument2,
      arg3: Argument3,
      arg4: Argument4
    ) => Promise<void>
  : [Argument6] extends [null]
  ? (
      store: ActionsStore<Store>
    ) => (
      arg1: Argument1,
      arg2: Argument2,
      arg3: Argument3,
      arg4: Argument4,
      arg5: Argument5
    ) => Promise<void>
  : [Argument7] extends [null]
  ? (
      store: ActionsStore<Store>
    ) => (
      arg1: Argument1,
      arg2: Argument2,
      arg3: Argument3,
      arg4: Argument4,
      arg5: Argument5,
      arg6: Argument6
    ) => Promise<void>
  : [Argument8] extends [null]
  ? (
      store: ActionsStore<Store>
    ) => (
      arg1: Argument1,
      arg2: Argument2,
      arg3: Argument3,
      arg4: Argument4,
      arg5: Argument5,
      arg6: Argument6,
      arg7: Argument7
    ) => Promise<void>
  : [Argument9] extends [null]
  ? (
      store: ActionsStore<Store>
    ) => (
      arg1: Argument1,
      arg2: Argument2,
      arg3: Argument3,
      arg4: Argument4,
      arg5: Argument5,
      arg6: Argument6,
      arg7: Argument7,
      arg8: Argument8
    ) => Promise<void>
  : [Argument10] extends [null]
  ? (
      store: ActionsStore<Store>
    ) => (
      arg1: Argument1,
      arg2: Argument2,
      arg3: Argument3,
      arg4: Argument4,
      arg5: Argument5,
      arg6: Argument6,
      arg7: Argument7,
      arg8: Argument8,
      arg9: Argument9
    ) => Promise<void>
  : (
      store: ActionsStore<Store>
    ) => (
      arg1: Argument1,
      arg2: Argument2,
      arg3: Argument3,
      arg4: Argument4,
      arg5: Argument5,
      arg6: Argument6,
      arg7: Argument7,
      arg8: Argument8,
      arg9: Argument9,
      arg10: Argument10
    ) => Promise<void>;

/**
 * The store passed to derived functions, containing the observable state and the libraries.
 *
 * @param Store
 * The store containing the state, actions and libraries.
 */
type DerivedStore<Store extends StoreType> = {
  state: ObservableState<Store>;
  libraries: Store["libraries"];
};

/**
 * A derived state function. It uses the store as the input of the first functions, the
 * rest of the inputs as the arguments of the second function and finally the last input as the returned
 * value of the function.
 * It's meant to be used when defining the derived state function.
 *
 * Examples:
 *  - `Derived<YourStore, string>` ---> `({ state }) => state.users[0].profile.name`
 *  - `Derived<YourStore, number, string>` ---> `({ state }) => id => state.users[id].profile.name`
 *  - `Derived<YourStore, number, string, string>` ---> `({ state }) => (id, prop) => state.users[id].profile[prop]`
 *
 * @param Store
 * The store containing both the state and the libraries.
 * @param ArgumentOrReturnX
 * The arguments of the derived function (if it has arguments). The last one is used as returned value.
 * Supports up to 10 arguments.
 * @param Return
 * The output if the derived function has arguments.
 */
export type Derived<
  Store extends StoreType,
  ArgumentOrReturn1,
  ArgumentOrReturn2 = null,
  ArgumentOrReturn3 = null,
  ArgumentOrReturn4 = null,
  ArgumentOrReturn5 = null,
  ArgumentOrReturn6 = null,
  ArgumentOrReturn7 = null,
  ArgumentOrReturn8 = null,
  ArgumentOrReturn9 = null,
  ArgumentOrReturn10 = null,
  Return = null
> = [ArgumentOrReturn2] extends [null]
  ? (store: DerivedStore<Store>) => ArgumentOrReturn1
  : [ArgumentOrReturn3] extends [null]
  ? (
      store: DerivedStore<Store>
    ) => (arg1: ArgumentOrReturn1) => ArgumentOrReturn2
  : [ArgumentOrReturn4] extends [null]
  ? (
      store: DerivedStore<Store>
    ) => (arg1: ArgumentOrReturn1, arg2: ArgumentOrReturn2) => ArgumentOrReturn3
  : [ArgumentOrReturn5] extends [null]
  ? (
      store: DerivedStore<Store>
    ) => (
      arg1: ArgumentOrReturn1,
      arg2: ArgumentOrReturn2,
      arg3: ArgumentOrReturn3
    ) => ArgumentOrReturn4
  : [ArgumentOrReturn6] extends [null]
  ? (
      store: DerivedStore<Store>
    ) => (
      arg1: ArgumentOrReturn1,
      arg2: ArgumentOrReturn2,
      arg3: ArgumentOrReturn3,
      arg4: ArgumentOrReturn4
    ) => ArgumentOrReturn5
  : [ArgumentOrReturn7] extends [null]
  ? (
      store: DerivedStore<Store>
    ) => (
      arg1: ArgumentOrReturn1,
      arg2: ArgumentOrReturn2,
      arg3: ArgumentOrReturn3,
      arg4: ArgumentOrReturn4,
      arg5: ArgumentOrReturn5
    ) => ArgumentOrReturn6
  : [ArgumentOrReturn8] extends [null]
  ? (
      store: DerivedStore<Store>
    ) => (
      arg1: ArgumentOrReturn1,
      arg2: ArgumentOrReturn2,
      arg3: ArgumentOrReturn3,
      arg4: ArgumentOrReturn4,
      arg5: ArgumentOrReturn5,
      arg6: ArgumentOrReturn6
    ) => ArgumentOrReturn7
  : [ArgumentOrReturn9] extends [null]
  ? (
      store: DerivedStore<Store>
    ) => (
      arg1: ArgumentOrReturn1,
      arg2: ArgumentOrReturn2,
      arg3: ArgumentOrReturn3,
      arg4: ArgumentOrReturn4,
      arg5: ArgumentOrReturn5,
      arg6: ArgumentOrReturn6,
      arg7: ArgumentOrReturn7
    ) => ArgumentOrReturn8
  : [ArgumentOrReturn10] extends [null]
  ? (
      store: DerivedStore<Store>
    ) => (
      arg1: ArgumentOrReturn1,
      arg2: ArgumentOrReturn2,
      arg3: ArgumentOrReturn3,
      arg4: ArgumentOrReturn4,
      arg5: ArgumentOrReturn5,
      arg6: ArgumentOrReturn6,
      arg7: ArgumentOrReturn7,
      arg8: ArgumentOrReturn8
    ) => ArgumentOrReturn9
  : [Return] extends [null]
  ? (
      store: DerivedStore<Store>
    ) => (
      arg1: ArgumentOrReturn1,
      arg2: ArgumentOrReturn2,
      arg3: ArgumentOrReturn3,
      arg4: ArgumentOrReturn4,
      arg5: ArgumentOrReturn5,
      arg6: ArgumentOrReturn6,
      arg7: ArgumentOrReturn7,
      arg8: ArgumentOrReturn8,
      arg9: ArgumentOrReturn9
    ) => ArgumentOrReturn10
  : (
      store: DerivedStore<Store>
    ) => (
      arg1: ArgumentOrReturn1,
      arg2: ArgumentOrReturn2,
      arg3: ArgumentOrReturn3,
      arg4: ArgumentOrReturn4,
      arg5: ArgumentOrReturn5,
      arg6: ArgumentOrReturn6,
      arg7: ArgumentOrReturn7,
      arg8: ArgumentOrReturn8,
      arg9: ArgumentOrReturn9,
      arg10: ArgumentOrReturn10
    ) => Return;

/**
 * Creates a new props object joining together the connected store and the normal props object.
 * It's meant to be used on components that are going to be connected.
 *
 * Examples:
 *  - `const Comp: React.FC<Connect<YourStore, { prop1: string }>> = ({ state, actions, libraries, prop1 }) => <div>...</div>`
 *  - `class Comp extends React.Component<Connect<YourStore, { prop1: string }>> { render() { return <div>...</div>; }}`
 *
 * @param Store
 * The store containing state, actions and libraries.
 * @param Props
 * The normal props object of this component.
 *
 */
export type Connect<Store extends StoreType, Props extends object = {}> = {
  state: ObservableState<Store>;
  actions: ExecutableActions<Store>;
  libraries: Store["libraries"];
} & Props;

/**
 * A helper to filter the properties of a store included in a bigger object, for example, when
 * the store is merged with React props.
 *
 * @param Store
 * The store containing state, actions and libraries.
 */
export type FilterStore<Store extends StoreType> = Omit<
  Store,
  "state" | "actions" | "libraries"
>;
