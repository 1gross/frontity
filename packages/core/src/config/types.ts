import { TransformOptions } from 'babel-core';

export type Mode = 'development' | 'production';
export type Env = 'module' | 'es5' | 'node';

export interface BabelConfig {
  module: TransformOptions;
  es5: TransformOptions;
  node: TransformOptions;
}
