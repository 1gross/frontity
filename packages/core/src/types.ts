import { TransformOptions } from "babel-core";
import { Configuration } from "webpack";

export type Mode = "development" | "production";
export type Target = "module" | "es5" | "node";

export interface Headers {
  [key: string]: string;
}

export type Robots = string;

export interface BabelConfigs {
  module: TransformOptions;
  es5: TransformOptions;
  node: TransformOptions;
}

export interface WebpackConfigs {
  module: Configuration;
  es5: Configuration;
  node: Configuration;
}

export interface HeaderConfig {
  static: Headers;
  dynamic: Headers;
}

export interface FrontityConfig {
  headers: HeaderConfig;
  robots: Robots;
  babel: BabelConfigs;
  webpack: WebpackConfigs;
}
