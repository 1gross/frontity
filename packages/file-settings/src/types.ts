// There are for types of settings:
// 1.a. Imported mono settings.
// 1.b. Imported multi settings.
// 2.a. Normalized mono settings.
// 2.b. Normalized multi settings.

type Base = {
  matches?: string[]; // Default: undefined
  mode?: string; // Default: "html"
  settings?: {
    url?: string; // Default: undefined
    title?: string; // Default: undefined
    timezone?: number; // Default: 0
    language?: string; // Default: "en"
  };
};

export type Package = {
  name: string; // Default: undefined
  active?: boolean; // Default: true
  namespaces?: string | string[]; // Default: undefined
  settings?: object; // Default: undefined
};

type Mono = Base & {
  name?: string;
};

type Multi = Base & {
  name: string;
};

type Imported<T> = Base & {
  packages: (string | T)[];
};

type Normalized<T> = Base & {
  packages: T[];
};

export type ImportedMono<T = Package> = Imported<T> & Mono;

export type ImportedMulti<T = Package> = Imported<T> & Multi;

export type NormalizedMono<T = Package> = Normalized<T> & Mono;

export type NormalizedMulti<T = Package> = Normalized<T> & Multi;

export type ImportedSettings<T = Package> =
  | ImportedMono<T>
  | ImportedMulti<T>[];

export type Settings<T = Package> = NormalizedMono<T> | NormalizedMulti<T>[];
