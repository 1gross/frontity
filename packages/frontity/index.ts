import create, { events } from "./src/commands/create";

events.on("create", console.log);
create({ name: "random-name" });
