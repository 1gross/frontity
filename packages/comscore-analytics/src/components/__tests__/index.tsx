/**
 * @jest-environment node
 */

import React from "react";
import TestRenderer from "react-test-renderer";
import { HelmetProvider } from "frontity";
import { State } from "frontity/types";
import { FilledContext } from "react-helmet-async";
import { Root as ComscoreAnalytics } from "..";
import ComscoreAnalyticsPkg from "../../../types";

const getState = (): State<ComscoreAnalyticsPkg> => ({
  analytics: { pageviews: { comscoreAnalytics: true } },
  comscoreAnalytics: {
    trackingIds: [],
  },
});

describe("Comscore Analytics", () => {
  test("should send if tracking ids are passed", () => {
    const state = getState();

    state.comscoreAnalytics.trackingIds = ["UA-XXXXXXX-X", "UA-YYYYYYY-Y"];

    const helmetContext = {};
    TestRenderer.create(
      <HelmetProvider context={helmetContext}>
        <ComscoreAnalytics state={state} actions={null} />
      </HelmetProvider>
    ).toJSON();
    const head = (helmetContext as FilledContext).helmet;

    expect(head.script.toString()).toMatchSnapshot();
  });

  test("should not send if no tracking ids are passed", () => {
    const state = getState();

    const helmetContext = {};
    TestRenderer.create(
      <HelmetProvider context={helmetContext}>
        <ComscoreAnalytics state={state} actions={null} />
      </HelmetProvider>
    ).toJSON();
    const head = (helmetContext as FilledContext).helmet;

    expect(head.script.toString()).toMatchSnapshot();
  });
});
