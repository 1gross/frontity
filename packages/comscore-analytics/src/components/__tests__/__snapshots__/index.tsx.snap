// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Comscore Analytics should not send if no tracking ids are passed 1`] = `""`;

exports[`Comscore Analytics should send if tracking ids are passed 1`] = `"<script data-rh=\\"true\\" async=\\"true\\" src=\\"https://sb.scorecardresearch.com/beacon.js\\"></script>"`;
